src/Mono.Android/Android.Runtime/AndroidRuntime.cs

```diff
@@ -388,52 +388,92 @@ public override void RegisterNativeMembers (JniType nativeClass, Type type, stri
 					return;
 				}
 
-				string[] members = methods!.Split ('\n');
-				if (members.Length < 2) {
+				int newLineIndex = methods!.IndexOf ('\n');
+				if (newLineIndex == -1) {
 					if (jniAddNativeMethodRegistrationAttributePresent)
 						base.RegisterNativeMembers (nativeClass, type, methods);
 					return;
 				}
 
-				JniNativeMethodRegistration[] natives = new JniNativeMethodRegistration [members.Length-1];
-				for (int i = 0; i < members.Length; ++i) {
-					string method = members [i];
-					if (string.IsNullOrEmpty (method))
-						continue;
-					string[] toks = members [i].Split (new[]{':'}, 4);
-					Delegate callback;
-					if (toks [2] == "__export__") {
-						var mname = toks [0].Substring (2);
-						MethodInfo? minfo = null;
-						foreach (var mi in type.GetMethods (BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.Static))
-							if (mi.Name == mname && JavaNativeTypeManager.GetJniSignature (mi) == toks [1]) {
-								minfo = mi;
-								break;
+				List<JniNativeMethodRegistration> natives = new List<JniNativeMethodRegistration> (16); // an initial capacity since most classes have more than 4-8 methods
+				ReadOnlySpan<char> methodsSpan = methods;
+
+				MethodInfo []? typeMethods = null;
+				while (!methodsSpan.IsEmpty) {
+					ReadOnlySpan<char> methodLine = methodsSpan.Slice (0, newLineIndex != -1 ? newLineIndex : methodsSpan.Length);
+					if (!methodLine.IsEmpty) {
+						SplitMethodLine (methodLine,
+							out ReadOnlySpan<char> name,
+							out ReadOnlySpan<char> signature,
+							out ReadOnlySpan<char> callbackString,
+							out ReadOnlySpan<char> callbackDeclaringTypeString);
+						Delegate callback;
+						if (callbackString.SequenceEqual ("__export__")) {
```

Comment:
What's the perf impact of using `.SequenceEqual()` here, particularly in terms of going through interfaces instead of concrete types (interface method dispatch is nominally slower)?

Reply:
The perf impact is identical to `==` on strings (albiet with a few special checks in `==` missing):

https://github.com/dotnet/runtime/blob/3c43ef6af7dd0a54e26ae23e4610f4937a7a7e4d/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L684-L709

Calls "EqualsHelper", which in turn just calls `SpanHelpers.SequenceEqual`:

https://github.com/dotnet/runtime/blob/3c43ef6af7dd0a54e26ae23e4610f4937a7a7e4d/src/libraries/System.Private.CoreLib/src/System/String.Comparison.cs#L24-L34

Where the SequenceEqual method being called here is:

https://github.com/dotnet/runtime/blob/3c43ef6af7dd0a54e26ae23e4610f4937a7a7e4d/src/libraries/System.Private.CoreLib/src/System/MemoryExtensions.cs#L456-L463

(note `IsBitwiseEquatable` is `true` for `char`).

There are no interfaces at work here. It is an extension method directly on `ReadOnlySpan<T>`.

Reply:
That's…wow: https://github.com/dotnet/runtime/blob/3c43ef6af7dd0a54e26ae23e4610f4937a7a7e4d/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs#L1152-L1241

Reply:
That's actually not the path it will go down for strings or `ReadOnlySpan<char>`. It will go down this path (taking the SIMD approach), since `IsBitwiseEquatable` is `true` for `char`:

https://github.com/dotnet/runtime/blob/3c43ef6af7dd0a54e26ae23e4610f4937a7a7e4d/src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Byte.cs#L1699-L1909